cf-agent/audit_mode.c

```diff
@@ -342,7 +345,104 @@ bool ManifestRename(const char *orig_name, const char *new_name)
     return true;
 }
 
-bool ManifestChangedFiles()
+static bool RunDiff(const char *path1, const char *path2)
+{
+    char diff_path[PATH_MAX];
+    strncpy(diff_path, GetBinDir(), sizeof(diff_path));
+    JoinPaths(diff_path, sizeof(diff_path), "diff");
+    char *command;
+    int ret = xasprintf(&command, "%s -u --label 'original %s' --label 'changed  %s' '%s' '%s'",
+                        diff_path, path1, path1, path1, path2);
+    assert(ret != -1); /* should never happen */
+
+    FILE *f = cf_popen(command, "r", true);
+
+    char buf[CF_BUFSIZE];
+    size_t n_read;
+    bool failure = false;
+    while (!failure && ((n_read = fread(buf, 1, sizeof(buf), f)) > 0))
+    {
+        /* TODO: fake the file paths? */
+        size_t to_write = n_read;
+        while (to_write > 0)
+        {
+            size_t n_written = fwrite(buf, 1, to_write, stdout);
+            if (n_written > 0)
+            {
+                to_write -= n_written;
+            }
+            else
+            {
+                Log(LOG_LEVEL_ERR, "Failed to print results from 'diff' for '%s' and '%s'",
+                    path1, path2);
+                failure = true;
+                break;
+            }
+        }
```

Comment:
I think there are problems with this while loop.

Here is the man page for fwrite:

```
     The functions fread() and fwrite() advance the file position indicator
     for the stream by the number of bytes read or written.  They return the
     number of objects read or written.  If an error occurs, or the end-of-
     file is reached, the return value is a short object count (or zero).

     The function fread() does not distinguish between end-of-file and error;
     callers must use feof(3) and ferror(3) to determine which occurred.  The
     function fwrite() returns a value less than nitems only if a write error
     has occurred.
```

At the very least, I think you need to advance `buf`, by creating and updating a temporary pointer / index, so that the retry tries to write from the right offset, in case of a short write.

Also, it seems fwrite should only return a short count in case of errors, is it better to log error then or retry?

Reply:
Yup, good catches! It's a code moved around so I'll add a new commit with a fix. I was actually reading the man page of `write()` and that's not so specific about the reasons for a failure.

Reply:
> Also, it seems fwrite should only return a short count in case of errors, is it better to log error then or retry?

Actually, I think it's better to retry and log an error if the retry doesn't write anything more.

Reply:
Makes sense 👍